clientv3/lease.go

```diff
@@ -40,3 +49,230 @@ type Lease interface {
 	// should be used instead of KeepAliveOnce.
 	KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error)
 }
+
+type lessor struct {
+	c *Client
+
+	mu     sync.Mutex       // guards all fields
+	conn   *grpc.ClientConn // conn in-use
+	remote pb.LeaseClient
+	stream pb.Lease_LeaseKeepAliveClient
+
+	keepAlives map[lease.LeaseID]chan *LeaseKeepAliveResponse
+	deadlines  map[lease.LeaseID]time.Time
+}
+
+func NewLease(c *Client) Lease {
+	l := &lessor{
+		c:    c,
+		conn: c.activeConnection(),
+
+		keepAlives: make(map[lease.LeaseID]chan *LeaseKeepAliveResponse),
+		deadlines:  make(map[lease.LeaseID]time.Time),
+	}
+
+	l.remote = pb.NewLeaseClient(l.conn)
+	return l
+}
+
+func (l *lessor) Create(ctx context.Context, ttl int64) (*LeaseCreateResponse, error) {
+	for {
+		r := &pb.LeaseCreateRequest{TTL: ttl}
+		resp, err := l.getRemote().LeaseCreate(ctx, r)
+		if err == nil {
+			return (*LeaseCreateResponse)(resp), nil
+		}
+
+		if isRPCError(err) {
+			return nil, err
+		}
+
+		if nerr := l.switchRemote(err); nerr != nil {
+			return nil, nerr
+		}
+	}
+}
+
+func (l *lessor) Revoke(ctx context.Context, id lease.LeaseID) (*LeaseRevokeResponse, error) {
+	for {
+		r := &pb.LeaseRevokeRequest{ID: int64(id)}
+		resp, err := l.getRemote().LeaseRevoke(ctx, r)
+		if err == nil {
+			return (*LeaseRevokeResponse)(resp), nil
+		}
+
+		if isRPCError(err) {
+			return nil, err
+		}
+
+		if nerr := l.switchRemote(err); nerr != nil {
+			return nil, nerr
+		}
+	}
+}
+
+func (l *lessor) KeepAlive(ctx context.Context, id lease.LeaseID) (chan<- *LeaseKeepAliveResponse, error) {
+	lc := make(chan *LeaseKeepAliveResponse, leaseResponseChSize)
+
+	// todo: add concellation based on the passed in ctx
+
+	l.mu.Lock()
+	_, ok := l.keepAlives[id]
+	if !ok {
+		l.keepAlives[id] = lc
+		l.deadlines[id] = time.Now()
+		l.mu.Unlock()
+		return lc, nil
+	}
+	l.mu.Unlock()
+
+	resp, err := l.KeepAliveOnce(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	lc <- resp
+	return lc, nil
+}
+
+func (l *lessor) KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
+	for {
+		resp, err := l.keepAliveOnce(ctx, id)
+		if err == nil {
+			return resp, err
+		}
+
+		nerr := l.switchRemote(err)
+		if nerr != nil {
+			return nil, nerr
+		}
+	}
+}
+
+func (l *lessor) keepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
```

Comment:
Got it. Thanks.

On Fri, Jan 29, 2016, 17:37 Xiang Li <notifications@github.com> wrote:

> In clientv3/lease.go
> <https://github.com/coreos/etcd/pull/4353#discussion_r51337133>:
>
> > +
> > +func (l *lessor) KeepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
> > +	for {
> > +		resp, err := l.keepAliveOnce(ctx, id)
> > +		if err == nil {
> > +			return resp, err
> > +		}
> > +
> > +		nerr := l.switchRemote(err)
> > +		if nerr != nil {
> > +			return nil, nerr
> > +		}
> > +	}
> > +}
> > +
> > +func (l *lessor) keepAliveOnce(ctx context.Context, id lease.LeaseID) (*LeaseKeepAliveResponse, error) {
>
> check the comment on the interface at line50
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/4353/files#r51337133>.
>
-- 
Sincerely,
Gyu-Ho Lee